### PR TITLE
[Tizen] Add basic example for NNstreamer for Tizen @open sesame 5/15 11:19

### PR DESCRIPTION
**Pull request by kparichay**
1. Verified loading of .so file provided by application by nnstreamer in Tizen
2. Run a basic application to run a pipeline with custom filter

tpk built using tizen studio, verified on Tizen 5.0 emulator

Resolves: nnsuite/nnstreamer#1139

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: Tizen.native/Application/inc/main.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://floralicense.org/license/
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
```
Comment (myungjoo) [inline Tizen.native/Application/inc/main.h]: We don't use FLORA license anymore. Please use Apache-2.

And it's 2019 :)

**Comment by myungjoo** [comment]
Question 1: Does this include ALL files required by Tizen Studio to build an app?

It is not recommended to have .so file in git repo. It is not portable (you cannot support other architectures) and you don't adapt changes in the code.

Thus, I recommend:
Approach 1: if RPM is used, add Requires to pull that .so file
Approach 2: if TPK is used, add a script to fetch that .so file from the built binary repo.

Note that you need to support both x86 and ARM.

**Comment by kparichay** [comment]
> Question 1: Does this include ALL files required by Tizen Studio to build an app?
I only included the code files now. I will update the PR to include all the files to build an app.

> It is not recommended to have .so file in git repo. It is not portable (you cannot support other architectures) and you don't adapt changes in the code.
> 
> Thus, I recommend:
> Approach 1: if RPM is used, add Requires to pull that .so file
> Approach 2: if TPK is used, add a script to fetch that .so file from the built binary repo.
> 
> Note that you need to support both x86 and ARM.
Thanks for the suggestion. I will create a script file to fetch the required .so files, and remove from the current PR.

**Comment by kparichay** [comment]
Updated - 
**V2:**
- Updated license to Apache 2.0
- Added extra files to directly import this folder as a Tizen Studio
project and build the app

**V3:**
- Download the custom filter so file from download.tizen.org

**V4:**
- Updated .gitignore

**Comment by myungjoo** [comment]
How did you solve the problem of rpm files incompatibility?

**Comment by kparichay** [comment]
> How did you solve the problem of rpm files incompatibility?

I downloaded all the nnstreamer rpm files from [http://download.tizen.org/snapshots/tizen/unified/latest/repos/$device/packages/i686/$rpmfile](download.tizen.org emulator packages for i686), along with the custom filter and it works.
Using nnstreamer rpm files from my own made rpm and custom filter from download.tizen.org did not work due to some compatibility issues.

**Comment by kparichay** [comment]
Verified working on latest tizen emulator image (with nnstreamer pre-installed)

File: Tizen.native/Application/inc/main.h (hunk 1 above)
Comment (kparichay) [reply Tizen.native/Application/inc/main.h]: Updated

**Comment by myungjoo** [comment]
@nnsuite/nnstreamer-committers Please build & test this commit in your (reviewer) own PC along with real Tizen machine (RPI3 or any devices with Tizen images). We need at least one review with actual Tizen app-dev experiences.

@kparichay There are quite a few people who has some experiences in Tizen Studio & Tizen devices :) let them experience and review your code

**Comment by kparichay** [comment]
> Please make it prepared for ARM builds. (at least in README)
> 
> Do you have ideas on how to generate build errors when an incorrect architecture is chosen or the so file is not downloaded?
> (e.g., try to "read" or "fetch" a symbol from the .so, which will enforce linker to check the so, which will fail if it does not exist or have incorrect arch.)

I will add a checker to verify the file with its architecture (by fetching a symbol from it).
I will update the README as well for the ARM builds in this PR.

**Comment by kparichay** [comment]
V2:
- Updated license to Apache 2.0
- Added extra files to directly import this folder as a Tizen Studio project and build the app
    
V3:
- Download the custom filter so file from download.tizen.org
    
V4:
- Updated .gitignore
    
V5:
- Added a checker to verify the working of so file
- Appended README for ARM builds

**Comment by leemgs** [comment]
To All reviewers, 
CPPcheck does not pass dlopen(3) usage. I will pass CPPcheck facility manually.

**Comment by myungjoo** [comment]
Have anyone tried to apply Tizen Studio to this? As we did with the previous Android app example, we need someone (other than the commiter) else to test build & run it.

**Comment by myungjoo** [comment]
It'd be perfect if we could upload the resulting .tpk to TIzen Store.
However, it won't be possible until we have Tizen 5.5 devices released.

**Pull request merged**